northstar/src/runtime/token.rs

```diff
@@ -4,197 +4,223 @@ use hmac::{
-    }
-
-    pub fn verify(&self, usage: &str) -> bool {
-        let hmac = Token::calculate_hmac(&self.time, usage);
-        let is_authentic = self.hmac == hmac; // constant time comparison ensured by CtOutput
-        let is_expired = Token::current_time()
-            .checked_sub(self.time)
-            .map_or(true, |age| age > TOKEN_EXPIRED_THRESHOLD);
-        is_authentic && !is_expired
-    }
-
-    pub fn write<T: std::io::Write>(&self, output: &mut T) -> std::io::Result<()> {
-        let time = self.time.to_be_bytes();
-        let hmac = self.hmac.clone().into_bytes();
-        output.write_all(&time)?;
-        output.write_all(&hmac)?;
-        Ok(())
+    /// Create a new token
+    pub fn new<U, T, S>(user: U, target: T, shared: S) -> Token
+    where
+        U: AsRef<[u8]>,
+        T: AsRef<[u8]>,
+        S: AsRef<[u8]>,
+    {
+        let now = now();
+        let hmac = calculate_hmac(&now, user.as_ref(), target.as_ref(), shared.as_ref());
+        Token { time: now, hmac }
+    }
+
+    /// Verify that `shared` matches the token
+    pub fn verify<U, T, S>(&self, user: U, target: T, shared: S) -> VerificationResult
+    where
+        U: AsRef<[u8]>,
+        T: AsRef<[u8]>,
+        S: AsRef<[u8]>,
+    {
+        let now = now();
+
+        if now < self.time {
+            VerificationResult::Future
+        } else if now - self.time > TOKEN_EXPIRED_THRESHOLD {
+            VerificationResult::Expired
+        } else if calculate_hmac(&self.time, user.as_ref(), target.as_ref(), shared.as_ref())
+            == self.hmac
+        {
+            VerificationResult::Ok
+        } else {
+            VerificationResult::Invalid
+        }
     }
+}
 
-    fn calculate_hmac(time: &Timestamp, usage: &str) -> Hmac {
-        let mut hasher = HmacSha256::new_from_slice(MAC_KEY.as_slice())
-            .expect("Failed to create SHA-256 hasher");
-        hasher.update(&time.to_be_bytes());
-        hasher.update(usage.as_bytes());
-        Hmac(hasher.finalize())
-    }
+#[cfg(not(test))]
+fn now() -> time::Duration {
+    time::Duration::from_secs(time::Instant::now().elapsed().as_secs())
+}
 
-    fn current_time() -> Timestamp {
-        std::time::SystemTime::now()
-            .duration_since(std::time::UNIX_EPOCH)
-            .expect("Clock is earlier than UNIX Epoch")
-            .as_millis()
-    }
+/// Special now impl for tests that need now in the past
+#[cfg(test)]
+fn now() -> time::Duration {
+    time::Duration::from_secs(
+        time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .unwrap()
+            .as_secs(),
+    )
 }
 
-#[derive(PartialEq)]
-struct Hmac(CtOutput<HmacSha256>);
+fn calculate_hmac(time: &time::Duration, user: &[u8], target: &[u8], shared: &[u8]) -> Hmac {
+    let mut hasher =
+        HmacSha256::new_from_slice(MAC_KEY.as_slice()).expect("Failed to create SHA-256 hasher");
+    hasher.update(&time.as_millis().to_be_bytes());
+    hasher.update(user);
```

Comment:
We should probably update the hasher with a delimiter character between the update calls. Otherwise, a token for a user with ID "abc" and a target with ID "xyz" is the same as one for the IDs "a" and "bcxyz".
Whatever delimiter character we chose (e.g. null byte), it should not be allowed to be part of a regular ID.

Reply:
Interesting. Today we allow anything that fulfils `AsRef<[u8]>` which can contain *anything* including a nul. We could take some (or all) bytes of the key?

Reply:
After thinking about this some more, I think we could HMAC both IDs individually and feed the outputs into the update calls. We can merge this first and I'll change it afterwards.

Reply:
Ah ok. Interesting. I added some parts of the key in between the two slices, but your approach sounds way cleaner.